java-squid/src/main/java/org/sonar/java/model/JavaTree.java

```diff
@@ -215,24 +205,100 @@ public void accept(TreeVisitor visitor) {
 
     @Override
     public Iterator<Tree> childrenIterator() {
+      Iterator<Tree> packageIterator = packageDeclaration == null ?
+        Iterators.<Tree>emptyIterator() :
+        Iterators.<Tree>singletonIterator(packageDeclaration);
       return Iterators.concat(
-        Iterators.singletonIterator(packageName),
+        packageIterator,
         imports.iterator(),
         types.iterator(),
-        packageAnnotations.iterator()
+        Iterators.singletonIterator(eofToken));
+    }
+
+    @Nullable
+    @Override
+    public PackageDeclarationTree packageDeclaration() {
+      return packageDeclaration;
+    }
+
+    @Override
+    public SyntaxToken eofToken() {
+      return eofToken;
+    }
+
+  }
+
+  public static class PackageDeclarationTreeImpl extends JavaTree implements PackageDeclarationTree {
+
+    private final List<AnnotationTree> annotations;
+    private final SyntaxToken packageKeyword;
+    private final ExpressionTree packageName;
+    private final SyntaxToken semicolonToken;
+
+    public PackageDeclarationTreeImpl(List<AnnotationTree> annotations, SyntaxToken packageKeyword, ExpressionTree packageName, SyntaxToken semicolonToken) {
+      super(Tree.Kind.PACKAGE);
+
+      this.annotations = Preconditions.checkNotNull(annotations);
+      this.packageKeyword = packageKeyword;
+      this.packageName = packageName;
+      this.semicolonToken = semicolonToken;
+
+      for (AnnotationTree annotationTree : annotations) {
+        addChild((AstNode) annotationTree);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) org/sonar/java/model/JavaTree$PackageDeclarationTreeImpl must not use com/sonar/sslr/api/AstNode [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AAstNode_SSLR_class_should_not_be_used)